spec.html

```diff
@@ -751,29 +751,38 @@ <h1>Algorithm Conventions</h1>
-      <p>When an algorithm is associated with a production alternative, the alternative is typically shown without any &ldquo;[ ]&rdquo; grammar annotations. Such annotations should only affect the syntactic recognition of the alternative and have no effect on the associated semantics for the alternative.</p>
-      <p>Syntax-directed operations are invoked with a parse node and, optionally, other parameters by using the conventions on steps 1, 3, and 4 in the following algorithm:</p>
-      <emu-alg>
-        1. Let _status_ be the result of performing SyntaxDirectedOperation of |SomeNonTerminal|.
-        2. Let _someParseNode_ be the parse of some source text.
-        2. Perform SyntaxDirectedOperation of _someParseNode_.
-        2. Perform SyntaxDirectedOperation of _someParseNode_ passing `"value"` as the argument.
-      </emu-alg>
-      <p>Unless explicitly specified otherwise, all chain productions have an implicit definition for every operation that might be applied to that production's left-hand side nonterminal. The implicit definition simply reapplies the same operation with the same parameters, if any, to the chain production's sole right-hand side nonterminal and then returns the result. For example, assume that some algorithm has a step of the form: &ldquo;Return the result of evaluating |Block|&rdquo; and that there is a production:</p>
-      <emu-grammar type="example">
-        Block :
-          `{` StatementList `}`
-      </emu-grammar>
-      <p>but the Evaluation operation does not associate an algorithm with that production. In that case, the Evaluation operation implicitly includes an association of the form:</p>
-      <p><b>Runtime Semantics: Evaluation</b></p>
-      <emu-grammar type="example">Block : `{` StatementList `}`</emu-grammar>
-      <emu-alg>
-        1. Return the result of evaluating |StatementList|.
-      </emu-alg>
+    <emu-clause id=sec-algorithm-conventions-directed-operations>
+      <h1>Directed Operations</h1>
+      <p>Some abstract operations, called <dfn>directed operations</dfn>, are implicitly parameterized with a target and have multiple definitions that may be selected depending on the value of the target. The target is referenced by callers of a directed operation using the method-like reference form `target.Operation()`. The target value may be referenced by the algorithm in a manner defined for that type of directed operation.</p>
```

Comment:
tbh using something like the bind operator might work better, like `target::Operation()`, if we had that available

Reply:
The spec is already using the syntax `target.Operation()`, and this PR generalizes its use (i.e., extends it in a compatible way).

Conversely, PR #1515 (BigInt) is proposing a `target::Operation()` syntax with different semantics from what you're suggesting. (In that PR, `target` is a *type* rather than a *value*.)

So why do you think that the `::` syntax might work better for this PR?

Reply:
I wasn’t making it as a concrete suggestion, since the bind operator isn’t a thing yet - more saying that using a dot, for a method that isn’t attached to the receiver, is confusingly non javascripty- so using an operator designed for this purpose seems nicer.

Reply:
I'm not really following you. Can you clarify what you mean by "attached to"? E.g., give cases where you think an abstract method is/isn't attached to a spec value?

Reply:
It’s entirely possible I’m misreading this PR. It seemed to me that a standalone function was being called with an object as the implicit receiver, but perhaps these aren’t standalone functions.

Reply:
They aren't functions at all, they're abstract operations. If that's what you meant, then I'm not sure what you mean by "standalone". Can you define it in spec terms, or give examples of standalone and non-standalone operations?

Reply:
An abstract operation is a standalone function - it isn't attached to any object. It doesn't have a receiver, just like a strict-mode non-bound JS function invoked with parens without using member access doesn't.

Reply:
So I think you're saying that you would would like to limit the spec notation:


To clarify, if `envRec` is some kind of environment record, and some alg step invokes `envRec.HasBinding(N)`, would you say that the `HasBinding` method:
- is standalone?
- is attached to `envRec`? 
- has a receiver?

Reply:
So I think you're saying that you'd prefer to limit the spec notation `X.Foo(args)` to cases where `Foo` is defined to be a component / attribute / "property" of `X`, and use something other that dot (e.g., `::`) when it isn't. This PR introduces notation like `|RelationalExpression|.Evaluate()`, but a syntax directed operation (like `Evaluate`) is not a component of a Parse Node, so you'd prefer the non-dot punctuation instead.

Reply:
Yes, that sounds right :-)

Reply:
Okay. So, applying that to existing constructs would go like this?:
- `_object_.[[InternalMethod]](...)` would stick with the dot (because the spec tends to talk about internal methods as if they're "attached" to the object).
- `_envRec_.EnvRecMethod(...)` would switch to the non-dot punctuation (because specific env record methods seem more like they belong to a particular env rec *type*).
- `_moduleRec_.ModuleRecMethod(...)` ditto EnvRecMethod